asciidoc/courses/llm-fundamentals/modules/2-vectors-semantic-search/lessons/2-vector-index/lesson.adoc

```diff
@@ -0,0 +1,222 @@
+
+
+== Vectorizing Movie Plots
+
+When you enrolled to this course, a Movie Recommendation Sandbox was created.
+This database consists of over 9000 movies, 15000 actors, and just over 100000 user ratings.
+
+Each movie has a `.plot` property.
+
+.Movie Plot Example
+[source,cypher]
+MATCH (m:Movie {title: "Toy Story"})
+RETURN m.title AS title, m.plot AS plot
+
+.Movie Plot Example Results
+[%cols="1"]
+|===
+| plot
+
+| "A cowboy doll is profoundly threatened and jealous when a new spaceman figure supplants him as top toy in a boy's room."
+|===
+
+You could manually search for other movies that mention the keywords in this plot, but that could be time-consuming.
+
+Instead, this text could be converted into an embedding and use the index in Neo4j to find the most similar movies.
+
+
+== Generating Embeddings
+
+We have prepared a file of 1000 embeddings created by calling the Ollama embedidngs endpoint, served by the link:https://github.com/docker/genai-stack[GenAI Stack^].
+The embeddings are created using the link:https://ollama.ai/library/orca-mini[Orca Mini] model.
+Orca Mini is a small, fast, LLM model popular with developers and data scientists because it still performs extremely well on minimal hardware.
+
+
+[TIP]
+.Generating the Embeddings
+====
+We wrote a simple Python script to call the `/api/embeddings` endpoint served by Ollama.
+You can link:TODO:link-the-repo[view the code on Github^].
+====
+
+For this lesson, it doesn't necessarily matter where these embeddings were obtained as each LLM will provide an embedding in its own shape.
+
+
+== Creating the Vector Index
+
+For simplicity, let's say the embeddings will be saved to the `.embedding` property on the `(:Movie)` node.
+We must create a vector index on that Label and property
+
+
+As vector indexes are a specialized type of index, you must call the `db.index.vector.createNodeIndex()` procedure.
+The procedure expects five parameters.
+
+.Procedure Signature
+[source,cypher]
+----
+CALL db.index.vector.createNodeIndex(
+    'moviePlots', // <1>
+    'Movie',      // <2>
+    'embedding',  // <3>
+    3200,         // <4>
+    'cosine'      // <5>
+)
+----
+
+1. The name of the index - we will want to call ours something `moviePlots`.
+2. The label on which to index - `Movie`.
+3. The key of the property on the node to index - `embedding`.
+4. The dimension of the embedding - Orca Mini embeddings consist of `3200` dimensions.
+5. The similarity function to use when comparing values in this index - this can be `euclidean` or `cosine`.
+
+[TIP]
+.Choosing a Similarity Function
+====
+You will want to choose the similarity function that is closest to the loss function used for training the embedding model.
+You may be able to find this out by reviewing your model provider's documentation.
+
+Experiment with the functions if necessary, and, if all else fails, experiment with `cosine` first.
```

Comment:
cosine is generally preferred for text embeddings as it is independent of the length of the vectors and only compares the angular distance